// .ignored/@cloudflare/puppeteer/lib/esm/puppeteer/puppeteer-core.js
/**
 * Copyright 2017 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { Puppeteer } from './common/Puppeteer.js';
import { WorkersWebSocketTransport } from './common/WorkersWebSocketTransport.js';
export * from './common/NetworkConditions.js';
export * from './common/QueryHandler.js';
export * from './common/DeviceDescriptors.js';
export * from './common/Errors.js';
class PuppeteerWorkers extends Puppeteer {
    constructor() {
        super({ isPuppeteerCore: true });
        this.connect = this.connect.bind(this);
        this.launch = this.launch.bind(this);
        this.sessions = this.sessions.bind(this);
        this.history = this.history.bind(this);
        this.limits = this.limits.bind(this);
    }
    /**
     * Launch a browser session.
     *
     * @param endpoint - Cloudflare worker binding
     * @returns a browser session or throws
     */
    async launch(endpoint, options) {
        let acquireUrl = '/v1/acquire';
        if (options === null || options === void 0 ? void 0 : options.keep_alive) {
            acquireUrl = `${acquireUrl}?keep_alive=${options.keep_alive}`;
        }
        const res = await endpoint.fetch(acquireUrl);
        const status = res.status;
        const text = await res.text();
        if (status !== 200) {
            throw new Error(`Unable to create new browser: code: ${status}: message: ${text}`);
        }
        // Got a 200, so response text is actually an AcquireResponse
        const response = JSON.parse(text);
        return this.connect(endpoint, response.sessionId);
    }
    /**
     * Returns active sessions
     *
     * @remarks
     * Sessions with a connnectionId already have a worker connection established
     *
     * @param endpoint - Cloudflare worker binding
     * @returns List of active sessions
     */
    async sessions(endpoint) {
        const res = await endpoint.fetch('/v1/sessions');
        const status = res.status;
        const text = await res.text();
        if (status !== 200) {
            throw new Error(`Unable to fetch new sessions: code: ${status}: message: ${text}`);
        }
        const data = JSON.parse(text);
        return data.sessions;
    }
    /**
     * Returns recent sessions (active and closed)
     *
     * @param endpoint - Cloudflare worker binding
     * @returns List of recent sessions (active and closed)
     */
    async history(endpoint) {
        const res = await endpoint.fetch('/v1/history');
        const status = res.status;
        const text = await res.text();
        if (status !== 200) {
            throw new Error(`Unable to fetch account history: code: ${status}: message: ${text}`);
        }
        const data = JSON.parse(text);
        return data.history;
    }
    /**
     * Returns current limits
     *
     * @param endpoint - Cloudflare worker binding
     * @returns current limits
     */
    async limits(endpoint) {
        const res = await endpoint.fetch('/v1/limits');
        const status = res.status;
        const text = await res.text();
        if (status !== 200) {
            throw new Error(`Unable to fetch account limits: code: ${status}: message: ${text}`);
        }
        const data = JSON.parse(text);
        return data;
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    async connect(endpointOrOptions, sessionId) {
        if (!sessionId) {
            return super.connect(endpointOrOptions);
        }
        const endpoint = endpointOrOptions;
        try {
            const transport = await WorkersWebSocketTransport.create(endpoint, sessionId);
            return super.connect({ transport, sessionId: sessionId });
        }
        catch (e) {
            throw new Error(`Unable to connect to existing session ${sessionId} (it may still be in use or not ready yet) - retry or launch a new browser: ${e}`);
        }
    }
}
const puppeteer = new PuppeteerWorkers();
export default puppeteer;
export const { connect, launch, sessions, history, limits } = puppeteer;
